base-trezor-keyring.js

```diff
@@ -0,0 +1,562 @@
+      hdPath: this.hdPath,
+      accounts: this.accounts,
+      page: this.page,
+      paths: this.paths,
+      perPage: this.perPage,
+      unlockedAccount: this.unlockedAccount,
+    });
+  }
+
+  deserialize(opts = {}) {
+    this.hdPath = opts.hdPath || hdPathString;
+    this.accounts = opts.accounts || [];
+    this.page = opts.page || 0;
+    this.perPage = opts.perPage || 5;
+    return Promise.resolve();
+  }
+
+  isUnlocked() {
+    return Boolean(this.hdk && this.hdk.publicKey);
+  }
+
+  unlock() {
+    if (this.isUnlocked()) {
+      return Promise.resolve('already unlocked');
+    }
+    return new Promise((resolve, reject) => {
+      this._getPublicKey({
+        path: this.hdPath,
+        coin: 'ETH',
+      })
+        .then((response) => {
+          if (response.success) {
+            this.hdk.publicKey = Buffer.from(response.payload.publicKey, 'hex');
+            this.hdk.chainCode = Buffer.from(response.payload.chainCode, 'hex');
+            resolve('just unlocked');
+          } else {
+            reject(
+              new Error(
+                (response.payload && response.payload.error) || 'Unknown error',
+              ),
+            );
+          }
+        })
+        .catch((e) => {
+          reject(new Error((e && e.toString()) || 'Unknown error'));
+        });
+    });
+  }
+
+  setAccountToUnlock(index) {
+    this.unlockedAccount = parseInt(index, 10);
+  }
+
+  addAccounts(n = 1) {
+    return new Promise((resolve, reject) => {
+      this.unlock()
+        .then((_) => {
+          const from = this.unlockedAccount;
+          const to = from + n;
+
+          for (let i = from; i < to; i++) {
+            const address = this._addressFromIndex(pathBase, i);
+            if (!this.accounts.includes(address)) {
+              this.accounts.push(address);
+            }
+            this.page = 0;
+          }
+          resolve(this.accounts);
+        })
+        .catch((e) => {
+          reject(e);
+        });
+    });
+  }
+
+  getFirstPage() {
+    this.page = 0;
+    return this.__getPage(1);
+  }
+
+  getNextPage() {
+    return this.__getPage(1);
+  }
+
+  getPreviousPage() {
+    return this.__getPage(-1);
+  }
+
+  __getPage(increment) {
+    this.page += increment;
+
+    if (this.page <= 0) {
+      this.page = 1;
+    }
+
+    return new Promise((resolve, reject) => {
+      this.unlock()
+        .then((_) => {
+          const from = (this.page - 1) * this.perPage;
+          const to = from + this.perPage;
+
+          const accounts = [];
+
+          for (let i = from; i < to; i++) {
+            const address = this._addressFromIndex(pathBase, i);
+            accounts.push({
+              address,
+              balance: null,
+              index: i,
+            });
+            this.paths[ethUtil.toChecksumAddress(address)] = i;
+          }
+          resolve(accounts);
+        })
+        .catch((e) => {
+          reject(e);
+        });
+    });
+  }
```

Comment:
I think `getPage` suggests giving it a page number? Using async primitives and starting index at 0 also makes it more clear, I think.

Roughly:

```suggestion
  __paginate(increment) {
    this.page = Math.max(0, this.page + increment);
    await this.unlock();
    const accounts = [];

    for (let i = 0; i < this.perPage; i++) {
      const index = this.page * this.perPage + i;
      const address = this._addressFromIndex(pathBase, index);
      accounts.push({
        address,
        balance: null,
        index,
      });
      this.paths[ethUtil.toChecksumAddress(address)] = index;
    }

    return accounts;
  }
```

Also, is there a significance wrt `_` vs `__` as prefix for private functions?

Reply:
I think we definitely should be looking into refactoring the pagination logic in a way such as the one you suggest. I was just wondering that, given how much of that code is almost identical within multiple keyrings (at least trezor and ledger), if it would be better to refactor pagination logic in a way that it can be reused.

Maybe it could be written within KeyringController and injected into individual keyrings, or maybe added to a new library along with other common tools for keyrings (e.g. hdkey for derivation).

I think we should assess if that's something we could be doing before refactoring individual keyrings. Any opinion on this @Gudahtt ?

Reply:
Reusing the pagination logic makes a lot of sense, but I don't think it would belong at the KeyringController level because not all keyrings have pagination.